File: C++/State/CellPhone/cell_phone_mode.cc
//
// Created by ericchen on 12/8/19.
//

#include <iostream>
#include "cell_phone_mode.h"

NormalMode::NormalMode(CellPhone *phone) : phone(phone) {}

void NormalMode::new_message() {
    std::cout << "Beep! " << phone->get_name() << "! You have a new message!\n";
}

void NormalMode::new_phone_call() {
    std::cout << "Ring ring! " << phone->get_name() << "! You are getting a new phone call!\n";
}

void NormalMode::new_app_notification() {
    std::cout << "Ding! " << phone->get_name() << "! You have a new app notification!\n";
}


SilentMode::SilentMode(CellPhone *phone) : phone(phone) {}

void SilentMode::new_message() {
    std::cout << "Vibrating! " << phone->get_name() << "! You have a new message!\n";
}

void SilentMode::new_phone_call() {
    std::cout << "Vibrating Vibrating!!! " << phone->get_name() << "! You are getting a new phone call!\n";
}

void SilentMode::new_app_notification() {
    std::cout << "Vibrating! " << phone->get_name() << "! You have a new app notification!\n";
}


AirplaneMode::AirplaneMode(CellPhone *phone) : phone(phone) {}

void AirplaneMode::new_message() {
    std::cout << "In airplane mode... Cannot receive any messages...\n";
}

void AirplaneMode::new_phone_call() {
    std::cout << "In airplane mode... Cannot receive any phone calls...\n";
}

void AirplaneMode::new_app_notification() {
    std::cout << "... " << phone->get_name() << "! You have a new app notification!\n";
}

File: Java/State/CellPhone/src/com/ericgtkb/CellPhoneMode.java
package com.ericgtkb;

public interface CellPhoneMode {
    void newMessage();
    void newPhoneCall();
    void newAppNotification();
}

class NormalMode implements CellPhoneMode {
    private CellPhone phone;

    public NormalMode(CellPhone phone) {
        this.phone = phone;
    }

    @Override
    public void newMessage() {
        System.out.println("Beep! " + phone.getName() + "! You have a new message!");
    }

    @Override
    public void newPhoneCall() {
        System.out.println("Ring ring! " + phone.getName() + "! You are getting a new phone call!");
    }

    @Override
    public void newAppNotification() {
        System.out.println("Ding! " + phone.getName() + "! You have a new app notification!");
    }
}

class SilentMode implements CellPhoneMode {
    private CellPhone phone;

    public SilentMode(CellPhone phone) {
        this.phone = phone;
    }

    @Override
    public void newMessage() {
        System.out.println("Vibrating! " + phone.getName() + "! You have a new message!");
    }

    @Override
    public void newPhoneCall() {
        System.out.println("Vibrating Vibrating!!! " + phone.getName() + "! You are getting a new phone call!");
    }

    @Override
    public void newAppNotification() {
        System.out.println("Vibrating! " + phone.getName() + "! You have a new app notification!");
    }
}

class AirplaneMode implements CellPhoneMode {
    private CellPhone phone;

    public AirplaneMode(CellPhone phone) {
        this.phone = phone;
    }

    @Override
    public void newMessage() {
        System.out.println("In airplane mode... Cannot receive any messages...");
    }

    @Override
    public void newPhoneCall() {
        System.out.println("In airplane mode... Cannot receive any phone calls...");
    }

    @Override
    public void newAppNotification() {
        System.out.println("... " + phone.getName() + "! You have a new app notification!");
    }
}